ipa-core/src/protocol/ipa_prf/boolean_ops/share_conversion_aby.rs

```diff
@@ -126,84 +128,113 @@ where
-            m_ctx.narrow(&Step::IntegerAddBetweenMasks),
-            record_id,
-            &sh_r,
-            &sh_s,
-        )
-        .await?;
-
-        // PRSS/Multiply masks added random highest order bit,
-        // remove them to not cause overflow in second addition (which is mod 256):
-        rs_with_higherorderbits[BITS - 1] = AdditiveShare::<Boolean, NC>::ZERO;
-
-        // return rs
-        rs_with_higherorderbits
-    };
-
-    // addition x+rs, where rs=r+s might cause carry
-    // this is not a problem since bit 255 of rs is set to 0
-    let (sh_y, _) = integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
-        m_ctx.narrow(&Step::IntegerAddMaskToX),
-        record_id,
-        &sh_rs,
-        &x,
-    )
-    .await?;
+    assert!(NC * input_shares.len() < 120_000);
+    // get context
+    let m_ctx = dzkp_validator.context();
+    // generate sub contexts
+    let ctx_gen = &m_ctx.narrow(&Step::GenerateSecretSharing);
+    let ctx_masks = &m_ctx.narrow(&Step::IntegerAddBetweenMasks);
+    let ctx_mask_x = &m_ctx.narrow(&Step::IntegerAddMaskToX);
+
+    // vector of intermediate results
+    let mut vec_sh_y =
+        Vec::<BitDecomposed<AdditiveShare<Boolean, NC>>>::with_capacity(input_shares.len());
+    let mut vec_sh_r =
+        Vec::<BitDecomposed<AdditiveShare<Boolean, NC>>>::with_capacity(input_shares.len());
+    let mut vec_sh_s =
+        Vec::<BitDecomposed<AdditiveShare<Boolean, NC>>>::with_capacity(input_shares.len());
+
+    for (i, x) in input_shares.iter().enumerate() {
+        let record_id_x = first_record_id + i;
+        // generate sh_r = (0, 0, sh_r) and sh_s = (sh_s, 0, 0)
+        // the two highest bits are set to 0 to allow carries for two additions
+        let (sh_r, sh_s) = gen_sh_r_and_sh_s::<_, BITS, NC>(ctx_gen, record_id_x);
+
+        // addition r+s might cause carry,
+        // this is no problem since we have set bit 254 of sh_r and sh_s to 0
+        let sh_rs = {
+            let (mut rs_with_higherorderbits, _) =
+                integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
+                    ctx_masks.clone(),
+                    record_id_x,
+                    &sh_r,
+                    &sh_s,
+                )
+                .await?;
+
+            // PRSS/Multiply masks added random highest order bit,
+            // remove them to not cause overflow in second addition (which is mod 256):
+            rs_with_higherorderbits[BITS - 1] = AdditiveShare::<Boolean, NC>::ZERO;
+
+            // return rs
+            rs_with_higherorderbits
+        };
+
+        // addition x+rs, where rs=r+s might cause carry
+        // this is not a problem since bit 255 of rs is set to 0
+        vec_sh_y.push(
+            integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
+                ctx_mask_x.clone(),
+                record_id_x,
+                &sh_rs,
+                x,
+            )
+            .await?
+            .0,
+        );
```

Comment:
I think it is worth keeping a separate assignment that grabs the sum, rather than `.0`, for clarity.

```suggestion
        let (sh_y, _) = integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
            ctx_mask_x.clone(),
            record_id_x,
            &sh_rs,
            x,
        )
        .await?;
        vec_sh_y.push(sh_y);
```